File: LeetCodeProblemsCMake/Problems/LeetCode2654MinimumNumberOfOperationsToMakeAllArrayElementsEqualTo1.cpp
//
// Created by saeed on 11/12/25.
//

#include "LeetCode2654MinimumNumberOfOperationsToMakeAllArrayElementsEqualTo1.h"

#include <iostream>
using namespace std;

int findGcd(int a, int b) {
    if (a % b == 0) return b;
    return findGcd(b, a % b);
}

int calcGcd(vector<int> &nums, int start, int end) {
    if (start == end)
        return nums[start];
    if (start - end == 1) {
        return findGcd(nums[start], nums[end]);
    }
    int mid = (start + end) >> 1;
    int leftGcd = calcGcd(nums, start, mid);
    int rightGcd = calcGcd(nums, mid + 1, end);
    return findGcd(leftGcd, rightGcd);
}

int LeetCode2654MinimumNumberOfOperationsToMakeAllArrayElementsEqualTo1::minOperations(vector<int> &nums) {
    int length = static_cast<int>(nums.size());
    int noneOneLength=length;
    for (int i=0;i<length;i++) {
        if (nums[i]==1)
            --noneOneLength;
    }
    if (noneOneLength!=length) {
        return noneOneLength;
    }
    for (int i=0;i<length;i++) {
        for (int j=0;j<length-i;j++) {
            if (calcGcd(nums,j,i+j)==1) {
                return length+i-1;
            }
        }
    }
    return -1;
}

void LeetCode2654MinimumNumberOfOperationsToMakeAllArrayElementsEqualTo1::Run() {
    vector<int> nums = {2, 6, 3, 4};
    cout << minOperations(nums) << '\n';
}

File: LeetCodeProblemsGo/LeetCode2654MinimumNumberOfOperationsToMakeAllArrayElementsEqualTo1/LeetCode2654MinimumNumberOfOperationsToMakeAllArrayElementsEqualTo1.go
package LeetCode2654MinimumNumberOfOperationsToMakeAllArrayElementsEqualTo1

func getGcd(a, b int) int {
	if (a % b) == 0 {
		return b
	}
	return getGcd(b, a%b)
}
func calcGcd(nums []int, memoisation [][]int, start, end int) int {
	if memoisation[start][end] != 0 {
		return memoisation[start][end]
	}
	if start == end {
		memoisation[start][end] = nums[start]
		return memoisation[start][end]
	}
	if end-start == 1 {
		memoisation[start][end] = getGcd(nums[start], nums[end])
		return memoisation[start][end]
	}
	mid := (start + end) >> 1
	memoisation[start][end] = getGcd(calcGcd(nums, memoisation, start, mid), calcGcd(nums, memoisation, mid+1, end))
	return memoisation[start][end]
}
func minOperations(nums []int) int {
	length := len(nums)
	noneZeroLength := length
	for _, num := range nums {
		if num == 1 {
			noneZeroLength--
		}
	}
	if noneZeroLength != length {
		return noneZeroLength
	}
	memoisation := make([][]int, length)
	for i := 0; i < length; i++ {
		memoisation[i] = make([]int, length)
	}
	for i := 0; i < length; i++ {
		for j := 0; j < length-i; j++ {
			if calcGcd(nums, memoisation, j, j+i) == 1 {
				return i + length - 1
			}
		}
	}
	return -1
}
